pkg/util/pki/csr_test.go

```diff
@@ -385,6 +386,218 @@ func TestRemoveDuplicates(t *testing.T) {
+		expectPathLenSet bool
+		expectedPathLen  int
+		expectError      bool
+	}{
+		"should not set MaxPathLen or MaxPathLenZero if unset on input": {
+			certificate: &cmapi.Certificate{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+				},
+				Spec: cmapi.CertificateSpec{
+					CommonName: "test-bundle-1",
+					IsCA:       true,
+				},
+			},
+			expectPathLenSet: false,
+		},
+		"should set MaxPathLen and MaxPathLenZero if input is zero": {
+			certificate: &cmapi.Certificate{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+				},
+				Spec: cmapi.CertificateSpec{
+					CommonName: "test-bundle-1",
+					MaxPathLen: makePathLen(0),
+					IsCA:       true,
+				},
+			},
+			expectPathLenSet: true,
+			expectedPathLen:  0,
+		},
+		"should set MaxPathLen but not MaxPathLenZero if input pathLen is greater than zero": {
+			certificate: &cmapi.Certificate{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+				},
+				Spec: cmapi.CertificateSpec{
+					CommonName: "test-bundle-1",
+					MaxPathLen: makePathLen(2),
+					IsCA:       true,
+				},
+			},
+			expectPathLenSet: true,
+			expectedPathLen:  2,
+		},
+		"should error if MaxPathLen set but IsCA is false": {
+			certificate: &cmapi.Certificate{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test",
+				},
+				Spec: cmapi.CertificateSpec{
+					CommonName: "test-bundle-1",
+					MaxPathLen: makePathLen(2),
+					IsCA:       false,
+				},
+			},
+			expectError: true,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			template, err := GenerateTemplate(test.certificate)
+			if (err != nil) != test.expectError {
+				t.Errorf("unexpected error from GenerateTemplate: %v", err)
+				return
+			}
+
+			if test.expectError {
+				return
+			}
+
+			if test.expectPathLenSet {
+				if template.MaxPathLen != test.expectedPathLen {
+					t.Errorf("wanted MaxPathLen to be '%d' but got '%d'", test.expectedPathLen, template.MaxPathLen)
+				}
+
+				if test.expectedPathLen == 0 && !template.MaxPathLenZero {
+					t.Error("expected pathLen of 0 but MaxPathLenZero is false")
+				} else if test.expectedPathLen != 0 && template.MaxPathLenZero {
+					t.Error("expected nonzero pathLen but MaxPathLenZero is true")
+				}
+			} else {
+				if template.MaxPathLen != 0 {
+					t.Errorf("expected MaxPathLen to be 0 but got: %d", template.MaxPathLen)
+				}
+
+				if template.MaxPathLenZero != false {
+					t.Error("expected MaxPathLenZero to be false but it was true")
+				}
+			}
+		})
+	}
+}
+
+var rawCSR = []byte(`-----BEGIN CERTIFICATE REQUEST-----
+MIH5MIGfAgEAMB8xHTAbBgNVBAMTFHRydXN0ZWQtaW50ZXJtZWRpYXRlMFkwEwYH
+KoZIzj0CAQYIKoZIzj0DAQcDQgAEa2wTLilxkRs5aF8EGvuTwlB1CdR13BTWzJg6
+vCmfbgg8dnG5CEaZ7Vt8vpNlAtJoP3LIB4mNScvzUKg9hoV1mKAeMBwGCSqGSIb3
+DQEJDjEPMA0wCwYDVR0PBAQDAgKkMAoGCCqGSM49BAMCA0kAMEYCIQCSdU3Qbpkl
+j3XqpAN2nbFTeAP8Qf2xHHZSlCKCqicT6wIhAPuTEMUe5GTuT00biauZjOaKXyRl
+jYssuLvyQHK2xoH2
+-----END CERTIFICATE REQUEST-----
+`)
```

Comment:
Not a huge fan of having random x509 data as test data hardcoded in- makes the test file more messy and is distracting imo :sweat_smile: . Can we extend and use the existing `testBundle` here?

Reply:
I marked as a draft in part because I was adding a comment to this explaining its purpose :sweat_smile: can change, I don't really feel strongly about it, but I'll push the reasoning in any case before I do change it